### DISCUSSION TITLE
Possible to upgrade server type without recreating the cluster?

### DISCUSSION THREAD

**Discussion by TheLoneDeveloper** [mysticaltech/terraform-hcloud-kube-hetzner#272] [Q&A] [answered]
Hello and thank you for this awesome tool.

We're currently running this in production with 3 control planes and 8 agents with server type of cpx21.
And we're soon going to need to upgrade/add servers.
So like the title says is it possible to upgrade the agents/control planes server type without recreating the entire cluster?

**Comment by mysticaltech** [comment, marked as answer]
Hey @TheLoneDeveloper, from the testing we have done, it should go without any issues, just increment the nodepool count for the one you want and apply, or even add a new nodepool (agent/control-plane all work the same). 

_But please do not hesitate to try before on a test cluster with more or less the same `kube.tf` file, as this is an open source project, and we try our best to try all relevant scenarios for each release but can't possibly test them all every time. (We are working on writing automated tests)._

**Reply by mysticaltech** [reply]
⚠️ @TheLoneDeveloper  **YES** it's possible, but I misread you at first. Here's how to do it:

if you want to upgrade the server type, create another nodepool of the kind you want (agent or control-plane), and add it to the end of the list. Apply.

When they are live and running, you must **drain** with kubectl the ones you want to decommission so that all workloads are automatically shifted to the "new" nodepool.

When this is done, and they are fully drained, only then can you can change the node count of the "old" nodepool to 0, then apply.

But DO NOT DELETE the "old" nodepool definitions (I know it's not pretty for the `kube.tf` file to have nodepools with counts of 0s, but that's how it works if you want to accomplish what you want), and do not change their order either, you always **append** to the list (or remove the last nodepool definition, that is valid too).

**Reply by TheLoneDeveloper** [reply]
Alright I will give this a try with a test cluster first and if it works i will proceed with upgrading the production cluster.

Thank you
